src/k2/cpo/service/CPOService.cpp

```diff
@@ -183,6 +195,184 @@ CPOService::handleGet(dto::CollectionGetRequest&& request) {
+        K2ERROR_EXC("Failed to push schema update: ", exc);
+        return seastar::make_ready_future<Status>(Statuses::S500_Internal_Server_Error("Failed to push schema"));
+    });
+}
+
+seastar::future<std::tuple<Status, dto::GetSchemasResponse>>
+CPOService::handleSchemasGet(dto::GetSchemasRequest&& request) {
+    auto it = schemas.find(request.collectionName);
+    if (it == schemas.end()) {
+        std::tuple<Status, dto::Collection> getResponse = _getCollection(request.collectionName);
+        if (!std::get<0>(getResponse).is2xxOK()) {
+            return RPCResponse(std::move(std::get<0>(getResponse)), dto::GetSchemasResponse{});
+        }
+
+        it = schemas.find(request.collectionName);
+        K2ASSERT(it != schemas.end(), "Schemas iterator is end after collection refresh");
+    }
+
+    return RPCResponse(Statuses::S200_OK("SchemasGet success"), dto::GetSchemasResponse { it->second });
+}
+
+seastar::future<std::tuple<Status, dto::CreateSchemaResponse>>
+CPOService::handleCreateSchema(dto::CreateSchemaRequest&& request) {
+    K2INFO("Received schema create request for " << request.collectionName << " : " << request.schema.name);
+
+    // 1. Stateless validation of request
+
+    Status validation = basicSchemaValidation(request.schema);
+    if (!validation.is2xxOK()) {
+        return RPCResponse(std::move(validation), dto::CreateSchemaResponse{});
+    }
+
+    // 2. Stateful validation
+
+    // getCollection to make sure in memory cache of schemas is up to date, and we need the 
+    // collection with partition map anyway to do the schema push
+    std::tuple<Status, dto::Collection> getResponse = _getCollection(request.collectionName);
+    if (!std::get<0>(getResponse).is2xxOK()) {
+        return RPCResponse(std::move(std::get<0>(getResponse)), dto::CreateSchemaResponse{});
+    }
+    dto::Collection& collection = std::get<1>(getResponse);
+
+    bool validatedKeys = false;
+    for (const dto::Schema& otherSchema : schemas[request.collectionName]) {
+        if (otherSchema.name == request.schema.name && otherSchema.version == request.schema.version) {
+            return RPCResponse(Statuses::S403_Forbidden("Schema name and version already exist"), dto::CreateSchemaResponse{});
+        }
+
+        // By induction, as long as we validate that the key fields match with at least one 
+        // other schema with the same name then they will always match
+        if (!validatedKeys && otherSchema.name == request.schema.name) {
+            validation = validateSchemaMatchingKeyDefs(otherSchema, request.schema);
+            if (!validation.is2xxOK()) {
+                return RPCResponse(std::move(validation), dto::CreateSchemaResponse{});
+            }
+
+            validatedKeys = true;
+        }
+    }
+
+
+    // 3. Update in memory
+
+    schemas[request.collectionName].push_back(std::move(request.schema));
+
+    // 4. Save to disk
+    Status saved = _saveSchemas(request.collectionName);
```

Comment:
if save to disk can fail, it maybe better to attempt to save first before updating in-memory state